stdlib/bytes.ml

```diff
@@ -167,7 +167,20 @@ let trim s =
   else
     empty
 
-let escaped s =
+let unsafe_escape ~share s =
+  (* We perform two passes on the input sequence, one to compute the
+     result size and one to write the result.
+
+     #11508, #11509: This logic would be incorrect in presence of
+     concurrent modification to the input, making the use of
+     [unsafe_set] below memory-unsafe.
+
+     Precondition: This function may be safely called only on
+     immutable or uniquely-owned byte sequences.
+
+     If [share] is true, then we may return the input byte sequence
+     itself when it contains no characters to escape. In all other
+     cases we return a fresh, uniquely-owned byte sequence. *)
```

Comment:
Unless I'm mistaken, you don't need the two `share` modes and can always do as if `share = false`:
- When called from Bytes.escape, `s` is a fresh copy of the input, uniquely owned, so it can be returned as the result.
- When called from String.escape, `s` is immutable and can be shared with the result.

Reply:
Good catch! The comment claims that we return a "fresh" sequence, but I can give the function a stronger specification (in the mutable case) where we consume the ownership of the input, and we return a byte sequence (fresh or not) with its ownership.